File: core/ucp/apps/ucp-api/src/adapters/outbound/identity/zitadel-organizations.adapter.ts
import { Inject, Injectable } from '@nestjs/common';
import { z } from 'zod';
import type { IOrganizationProvider } from '../../../ports/outbound/organization.provider.js';
import type { IProjectProvider } from '../../../ports/outbound/project.provider.js';
import { PROJECT_PROVIDER } from '../../../ports/outbound/project.provider.js';
import { ZitadelBaseClient } from './zitadel-base.client.js';

const CreateOrgResponseSchema = z.object({
  id: z.string().optional(),
  organizationId: z.string().optional(),
  orgId: z.string().optional(),
});

@Injectable()
export class ZitadelOrganizationsAdapter
  extends ZitadelBaseClient
  implements IOrganizationProvider
{
  constructor(
    @Inject(PROJECT_PROVIDER)
    private readonly projectProvider: IProjectProvider,
  ) {
    super();
  }

  async createOrganization(
    name: string,
  ): Promise<{ orgId: string; grantSucceeded?: boolean }> {
    this.logger.log(`Provisioning Organization in Zitadel: ${name}`);

    try {
      const response = await this.fetchWithAuth('/management/v1/orgs', {
        method: 'POST',
        body: JSON.stringify({ name }),
      });

      if (!response.ok) {
        await this.handleResponseError(response, 'create org');
      }

      const data = (await response.json()) as unknown;
      const parsedData = CreateOrgResponseSchema.parse(data);
      const orgId =
        parsedData.id || parsedData.organizationId || parsedData.orgId;
      if (!orgId) throw new Error('Org ID not returned from Zitadel');
      this.logger.log(`Created Organization in Zitadel with ID: ${orgId}`);

      let grantSucceeded = false;
      if (this.ucpProjectId) {
        // Enterprise Grade: Dynamically query the source of truth (Zitadel Project)
        // for all roles assigned to the configured tenant group so we never have to hardcode role additions.
        const tenantGroup = process.env.ZITADEL_TENANT_ROLE_GROUP || 'Tenant';
        const allRoles = await this.projectProvider.getRoles();
        const tenantRoleKeys = allRoles
          .filter((role) => role.group === tenantGroup) // Groups are configured in main.tf
          .map((role) => role.key);

        try {
          await this.projectProvider.createProjectGrant(
            orgId,
            this.ucpProjectId,
            tenantRoleKeys,
          );
          grantSucceeded = true;
        } catch (error) {
          this.logger.error(
            `Failed to grant UCP project to org ${orgId}. The organization was created successfully but project grant failed. Manual intervention or retry may be required.`,
            error,
          );
          // Don't throw, let org creation succeed even if grant fails
          // Caller should check grantSucceeded and handle accordingly
        }
      }

      return { orgId, grantSucceeded };
    } catch (error) {
      this.logger.error('Error creating organization in Zitadel', error);
      throw error;
    }
  }

  async deleteOrganization(orgId: string): Promise<void> {
    this.logger.log(`Deleting Organization in Zitadel: ${orgId}`);

    try {
      // First try admin v1 delete (which works cross-org)
      let response = await this.fetchWithAuth(`/admin/v1/orgs/${orgId}`, {
        method: 'DELETE',
      });

      if (!response.ok) {
        // Fallback to management v1 if admin fails
        response = await this.fetchWithAuth(`/management/v1/orgs/${orgId}`, {
          method: 'DELETE',
        });
      }

      if (!response.ok) {
        await this.handleResponseError(response, 'delete org');
      }

      this.logger.log(
        `Successfully deleted Organization ${orgId} from Zitadel`,
      );
    } catch (error) {
      this.logger.error(
        `Error deleting organization ${orgId} in Zitadel`,
        error,
      );
      throw error;
    }
  }
}

File: core/ucp/apps/ucp-api/src/ucp_api/adapters/outbound/identity/zitadel_organizations_adapter.py
import logging
from typing import Tuple
from ucp_api.adapters.outbound.identity.zitadel_client import ZitadelClient
from ucp_api.ports.outbound.organization_provider import IOrganizationProvider
from ucp_api.ports.outbound.project_provider import IProjectProvider
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)


class CreateOrgResponse(BaseModel):
    id: str | None = None
    organization_id: str | None = Field(None, alias="organizationId")
    org_id: str | None = Field(None, alias="orgId")


class ZitadelOrganizationsAdapter(ZitadelClient, IOrganizationProvider):
    def __init__(self, project_provider: IProjectProvider) -> None:
        super().__init__()
        self.project_provider = project_provider

    async def create_organization(self, name: str) -> Tuple[str, bool]:
        logger.info(f"Provisioning Organization in Zitadel: {name}")

        try:
            response = await self.fetch_with_auth(
                endpoint="/management/v1/orgs", method="POST", json={"name": name}
            )

            if response.status_code >= 400:
                await self.handle_response_error(response, "create org")

            data = response.json()
            parsed_data = CreateOrgResponse.model_validate(data)
            org_id = parsed_data.id or parsed_data.organization_id or parsed_data.org_id

            if not org_id:
                raise ValueError("Org ID not returned from Zitadel")

            logger.info(f"Created Organization in Zitadel with ID: {org_id}")

            grant_succeeded = False
            if self.ucp_project_id:
                tenant_group = self.settings.zitadel_tenant_role_group
                all_roles = await self.project_provider.get_roles()
                tenant_role_keys = [role.key for role in all_roles if role.group == tenant_group]

                try:
                    await self.project_provider.create_project_grant(
                        org_id, self.ucp_project_id, tenant_role_keys
                    )
                    grant_succeeded = True
                except Exception as error:
                    logger.error(
                        f"Failed to grant UCP project to org {org_id}. "
                        "The organization was created successfully but project grant failed. "
                        f"Manual intervention or retry may be required: {error}"
                    )

            return org_id, grant_succeeded
        except Exception as error:
            logger.error(f"Error creating organization in Zitadel: {error}")
            raise error

    async def delete_organization(self, org_id: str) -> None:
        logger.info(f"Deleting Organization in Zitadel: {org_id}")

        try:
            # First try admin v1 delete (which works cross-org)
            response = await self.fetch_with_auth(
                endpoint=f"/admin/v1/orgs/{org_id}", method="DELETE"
            )

            if response.status_code >= 400:
                # Fallback to management v1 if admin fails
                response = await self.fetch_with_auth(
                    endpoint=f"/management/v1/orgs/{org_id}", method="DELETE"
                )

            if response.status_code >= 400:
                await self.handle_response_error(response, "delete org")

            logger.info(f"Successfully deleted Organization {org_id} from Zitadel")
        except Exception as error:
            logger.error(f"Error deleting organization {org_id} in Zitadel: {error}")
            raise error
